server/src/main/java/org/elasticsearch/index/search/QueryStringQueryParser.java

```diff
@@ -789,6 +798,8 @@ public Query parse(String query) throws ParseException {
         if (query.trim().isEmpty()) {
             return Queries.newMatchNoDocsQuery("Matching no documents because no terms present");
         }
-        return super.parse(query);
+        Query result = super.parse(query);
+        checkForTooManyFields(this.queriedFields.size(), this.context);
```

Comment:
This is not enough, we should check **every** field expansion not just the last one (when `extractMultiFields` is called).

Reply:
My hope with putting this check later here was that we can report something close to the real number of fields used in the query to the user with the error message. I had the check in `extractMultiFields` earlier but then we throw an error whenever the parsing process exceeds the limit. I was looking for a way to get a mor accurate could of the total number of used fields, but I see now that this is too late here. Need to rethink this a bit and adapt tests.

Reply:
If we do that, I'd prefer to do it in a follow up since that would be breaking (we don't check the sum of unique fields at the moment). Lucene 9 will also check the number of clauses globally so this is probably not needed to implement such checks at this level.

Reply:
Agreed, if there are other clause checks internal to Lucene this isn't needed. I thought it was achievable but see how thats not the case now.